File: codes/Garnetwzy/78.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ret;
        if(nums.size() == 0)
            return ret;
        vector<bool> visit(nums.size(), false);
        vector<int> cur = {};
        for(int i = 0; i <= nums.size(); i++) {
            dfs(ret, cur, nums, visit, i, 0);
        }
        return ret;
    }
    
    void dfs(vector<vector<int>>& ret, vector<int>& cur, vector<int>& nums, vector<bool>& visit, int len, int index) {
        if(cur.size() == len){
            ret.push_back(cur);
            return;
        }
        for(int i = index; i < nums.size(); i++) {
            if(visit[i])
                continue;
            visit[i] = true;
            cur.push_back(nums[i]);
            dfs(ret, cur, nums, visit, len, i+1);
            cur.pop_back();
            visit[i] = false;
        }
    }
};

File: codes/HellDream/78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> lists = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        Arrays.sort(nums);
        generateSubsets(lists, list, nums,0);
        return lists;
    }
    
    public void generateSubsets(List<List<Integer>> lists, List<Integer> list, int[] nums,int index){
        if(!lists.contains(list))
            lists.add(list);
        for(int i=index;i<nums.length;i++){
            List<Integer> newList = new ArrayList<>(list);
            newList.add(nums[i]);
            generateSubsets(lists, newList, nums,i+1);
        }
    }
}